New subport: gromacs-plumed

GiovanniBussi opened the pull request:
This subport allows installing gromacs with plumed support.

###### Description

I enclose a pull request for a modified version of gromacs which includes the PLUMED library. Could anyone check it and, in case is ok, merge it? See also explanations below.

**How does GROMACS+PLUMED work.** [PLUMED](http://www.plumed.org) provides a library and a command line tool. It can be currently installed with MacPorts (`port install plumed`). The library is used already in some molecular dynamics code. However, for some other codes, including gromacs, a patch is necessary. The patch is contained in PLUMED itself and is applied with a command such as `plumed patch -p`.

**Variant vs subport**. I originally implemented a variant `gromacs +plumed`. However, based on the feedbacks I received during the review of #408, I now think it is better to have a subport `gromacs-plumed`. I make the subport conflict with `gromacs` so that either one or the other is installed. I am happy to change it back to a variant in case MacPorts developers think it is better in this case (it is a two lines change).

**Portfile changes**. The only changes are those required for the `gromacs-plumed` subport. In particular, before installing gromacs, the patch is applied using the `plumed patch` command.

Please let me know if this request is ready for merge or if I should make some change. Thanks!

Giovanni
 
###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->
- [ ] bugfix
- [X] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.11.6 15G1611
Xcode 8.1 8B62 

###### Verification <!-- (delete not applicable items) -->
Have you
- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

dstrubbe commented:
Hello, regarding the subport vs variant question, I would say -- unless the plumed patches significantly modify or compromise the normal functionality of gromacs, I don't really a see a reason to use a subport rather than a variant. Is there a reason someone would want to have both patched and unpatched gromacs installed? I am not sure what analogy you are finding to the discussion on your plumed port submission.

GiovanniBussi commented:
Dear @dstrubbe

thanks for your comment. In principle plumed patch should not alter gromacs behavior. In any case, the way I wrote the subport (with `conflicts gromacs`) it would not be possible to install both `gromacs` and `gromacs-plumed`. Indeed, they both install an executable named `gmx`.

The reason why I used the subport was indeed a different one, that is to allow pre-compiled binaries to be distributed by macports servers.
````
 sudo port install gromacs +plumed
````
would always require full recompilation of gromacs, whereas
````
 sudo port install gromacs-plumed
````
could download a pre-compiled binary. I would prefer it due to the long time required to compile gromacs.

Using subports in order to allow binary distribution was suggested by @mojca reviewing #408, although in a different context (i.e., ports that could be installed simultaneously). However, I noticed in the port tree that many Portfiles use a subport in order to distribute alternative versions (e.g. `*-devel` versions), so I thought the idea can be applied here.

Notice that this benefit disappears anyway when you use non-default variants (e.g. `+gcc6`), because only default variants are pre-compiled.

Anyway, the variant version works as well, so if you prefer I am happy to change back to a variant `+plumed`.

Giovanni

GiovanniBussi commented:
Dear @dstrubbe 

could you please check if the pull request can be merged?

In case you prefer I am happy to change it to a `+plumed` variant  (in order to do `port install gromacs +plumed`), though I would prefer the subport for the reason explained above (precompiled binaries).

Notice that there are now two plumed ports (`plumed` and `plumed-devel`, the latter being a subport with a development version). They way I set the dependencies (on files `prefix/bin/plumed` and `prefix/lib/libplumedKernel.dylib`) should allow both plumed ports to satisfy the dependency.

Thanks!

Giovanni

dstrubbe commented:
Ok having the subport for the reasons you mentioned seems fine. I see you remove a compatibility check of the MPI used for gromacs and plumed. If an incompatibility would indeed cause problems when linked, then I think we should have the check. I understand the issue of plumed vs plumed-devel, but you can solve that by running the plumed check if it is installed and otherwise the plumed-devel check.

GiovanniBussi commented:
@dstrubbe good point...
I will fix this then let you know. Thanks!
Giovanni

GiovanniBussi commented:
@dstrubbe I think I fixed the issue. In particular, if mpi variant is set, it makes sure the plumed has the same variant.

In addition, I added a command to disable thread-MPI (the internal implementation of MPI available with gromacs) when using plumed, since plumed only works correctly with a real MPI library.

Let me know if there is something else I should do for the port to be merged. Thanks!

Giovanni

dstrubbe commented:
Ok, looks fine to me.

GiovanniBussi commented:
@dstrubbe if you are ok with the change, could you please merge it?

In case you do not have permissions, perhaps you have to mark the review as passed so that some MacPorts developer can merge it.

If necessary, let me know if I should do further modifications. Thanks a lot!

Giovanni
